sprint_06

h8ax opened the pull request:
Приложение запускается и работает корректно по функциональным и техническим требованиям.

Добавлена работа с сетью (фильмы для квиза должны загружаться из открытого API IMDb).

В приложении использованы фильмы из топ-250 и самых популярных фильмов IMDb.

Обновлён метод viewDidLoad() в MovieQuizViewController — там вы должны начинать загрузку данных и показывать индикатор загрузки (уже показанный метод showLoadingIndicator()).

Если происходит ошибка загрузки данных о фильмах, то показывается алерт с ошибкой.

Приложение не пишет в логи сообщения об ошибках, связанных с работой с UI из другого потока.

h8ax committed:
sprint_06

Helvetii951 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -9,118 +16,181 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate
     
     // MARK: - QuestionFactoryDelegate
     
+    private func showLoadingIndicator() {
+        activityIndicator.isHidden = false
+        activityIndicator.startAnimating()
+    }
```
Рекомендую посмотреть на свойство [hidesWhenStopped](https://developer.apple.com/documentation/uikit/uiactivityindicatorview/1622837-hideswhenstopped) 
у `activityIndicator`, и использовать его вместо ручного управления свойством `isHidden`. 

Когда ты будешь вызывать `startAnimating` – индикатор будет появляться автоматически, соответственно на `stopAnimating` – скрываться, а свойство `isHidden` больше не понадобится, индикатор все сделает за тебя. 

**Важно:** это свойство настраивают всего раз раз: кодом во `viewDidLoad()` или аналогичной галочкой у элемента на сториборде.
А строки `activityIndicator.isHidden = false(true)` нужно будет убрать

Helvetii951 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -9,118 +16,181 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate
+    }
+    
+    func didLoadDataFromServer() {
+        activityIndicator.isHidden = true
```
Рекомендую сделать функцию `hideLoadingIndicator` по типу `showLoadingIndicator`.

Helvetii951 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -9,118 +16,181 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate
+    
+    func didLoadDataFromServer() {
+        activityIndicator.isHidden = true
+        questionFactory?.requestNextQuestion()
```
Ты можешь дополнительно показывать `activityIndicator` между загрузкой вопросов, потому что загрузка вопроса тоже занимает некоторое время из-за получения изображения.
Перед запросом следующего вопроса – показывать индикатор, а в момент получения – выключать анимацию.

Helvetii951 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -4,30 +4,71 @@
+                
+                var imageData = Data()
+                
+                do {
+                    imageData = try Data(contentsOf: movie.resizedImageURL)
+                } catch {
+                    print("Failed to load image")
+                }
```
Обращу внимание, что игра состоит из вопроса о рейтинге, на который отвечает пользователь, и контекста о фильме в виде постера(картинка). 
Без постера, пользователю будет затруднительно ответить на вопрос о рейтинге фильма верно.

В этом месте, функция не завершает свою работу и идет формировать вопрос без постера. 
В таком случае, бизнес-логика приложения будет нарушена, потому что пользователь не сможет ответить корректно, ведь изображения нет.
Это состояние стоит обработать, например, с помощью алерта. 
Не забудь, что алерт будет обновлять UI и это тоже нужно учесть (вернуть на main), ровно также, как и прерывание функции (сделать return)

Если зачхочешь это реализовать, обрати внимание, что ошибку загрузки изначальных данных и ошибку подгрузки вопроса нужно обрабатывать отдельно.

Helvetii951 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -4,30 +4,71 @@
+                let raiting = Float(movie.rating) ?? 0
+                
+                let text = " Рейтинг этого фильма больше, чем 7?"
+                let correctAnswer = raiting > 7
```
Я бы порекомендовал чуть разнообразить гэймплэй. Не просто сравнивать с 7, а брать предполагаемое значение из интервала 6...9.

Helvetii951 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -4,30 +4,71 @@
-         self.delegate = delegate
-     }
+    private let moviesLoader: MoviesLoading
+    private var delegate: QuestionFactoryDelegate?
```
Обычно ссылки на делегаты делают weak. И именно здесь у тебя возник `Retain cycle`.

h8ax committed:
реализовано hidesWhenStopped
показан activityIndicator между вопросами
переделан alert
разнообразил геймплей
убрал retainCycle

Helvetii951 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -4,30 +4,78 @@
+            guard let self = self else { return }
+            
+            let index = (0..<self.movies.count).randomElement() ?? 0
+            guard let movie = self.movies[safe: index] else {
+                DispatchQueue.main.async { [weak self] in
+                    guard let self = self else { return }
+                    self.delegate?.didFailToLoadData(with: NSError(domain: "QuestionFactory", code: 0, userInfo: [NSLocalizedDescriptionKey: "Failed to load movie data"]))
+                }
+                return
+            }
```
Здесь пользователь получит ошибку, но повторную попытку загрузки фильмов совершить не удастся. Т.е. действия кода будет такие:
1. Вызываем `questionFactory?.loadData()` во `viewDidLoad`
2. При ошибке загрузки вызываем в алерте `self.questionFactory?.requestNextQuestion()`
3. Натыкаемся на `guard let movie = self.movies[safe: index]`. Показываем алерт.
4. Вызываем в алерте `self.questionFactory?.requestNextQuestion()`
5. Натыкаемся на `guard let movie = self.movies[safe: index]`. Показываем алерт.
6. И так мы зациклимся. Данные так и не подгрузятся.


Получение данных и загрузка контента для вопроса - разные ошибки. 

- При получении ошибки загрузки фильмов на старте приложения нужно показать алерт с попыткой `questionFactory?.loadData()`. 
- При ошибке подгрузки  алерт с `self.questionFactory?.requestNextQuestion()`

Можно добавить в делегат колбэк didFailToLoadImage/Question, и вызывать при обработке ошибки загрузки вопроса.

Вот этот код лучше убрать и вернуть, как было: 
```
guard let movie = self.movies[safe: index] else {
    DispatchQueue.main.async { [weak self] in
         guard let self = self else { return }
         self.delegate?.didFailToLoadData(with: NSError(domain: "QuestionFactory", code: 0, userInfo: [NSLocalizedDescriptionKey: "Failed to load movie data"]))}
     return
}
```

h8ax committed:
Поправил showNetworkError
